testcases/crypto/aes_func.c

```diff
@@ -1929,6 +2006,308 @@ CK_RV do_SignVerifyMAC(struct published_mac_test_suite_info *tsuite)
+        {CKA_EXTRACTABLE, &bfalse, sizeof(CK_BBOOL)},
+        {CKA_VALUE_LEN, &keylen, sizeof(CK_ULONG)}
+    };
+    CK_ULONG keygen_tmpl_len = sizeof(keygen_tmpl) / sizeof(CK_ATTRIBUTE);
+
+    CK_ATTRIBUTE find_tmpl[] = {
+        {CKA_LABEL, &label, sizeof(label)},
+    };
+    CK_ULONG find_tmpl_len = sizeof(find_tmpl) / sizeof(CK_ATTRIBUTE);
+    CK_ULONG count = 0;
+
+    CK_BYTE input[] =
+        { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+          0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f };
+    CK_BYTE output[32];
+    CK_BYTE decrypted[32];
+    CK_ULONG outlen, decrlen;
+
+    CK_BYTE user_pin[PKCS11_MAX_PIN_LEN];
+    CK_FLAGS flags;
+    CK_ULONG user_pin_len;
+    CK_SLOT_ID slot_id = SLOT_ID;
+
+    testsuite_begin("do_EncryptDecryptAESPkey");
+    testcase_rw_session();
+    testcase_user_login();
+
+    /* Skip tests if the slot doesn't support protected keys*/
+    if (!is_ep11_token(slot_id)) {
+        testsuite_skip(6, "Slot %u doesn't support protected keys",
+                       (unsigned int) slot_id);
+        goto testcase_cleanup;
+    }
+
+    testcase_begin("Generate token key object and encrypt");
+    testcase_new_assertion();
+
+    /* Generate token object */
+    rc = funcs->C_GenerateKey(session, &keygen_mech, keygen_tmpl, keygen_tmpl_len, &h_key);
+    if (rc != CKR_OK) {
+        testcase_fail("C_GenerateKey rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    /* Encrypt some data with this key object */
+    rc = funcs->C_EncryptInit(session, &encr_mech, h_key);
+    if (rc != CKR_OK) {
+        testcase_fail("C_EncryptInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    memset(output, 0, sizeof(output));
+    rc = funcs->C_Encrypt(session, input, sizeof(input), output, &outlen);
+    if (rc != CKR_OK) {
+        testcase_fail("C_Encrypt rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    testcase_pass("Generate token key object and encrypt");
+    testcase_begin("Log out, close session, and login in again in new session");
+    testcase_new_assertion();
+
+    /* Log out and close session */
+    testcase_user_logout();
+    rc = funcs->C_CloseAllSessions(slot_id);
+    if (rc != CKR_OK) {
+        testcase_fail("C_CloseAllSessions rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    /* Log in again read/only, ok for just decrypt with an existing pkey */
+    testcase_rw_session();
+    testcase_user_login();
+
+    testcase_pass("Log out, close session, and login in again in new session");
+    testcase_begin("Retrieve the key from repository");
+    testcase_new_assertion();
+
+    /* Retrieve the key from repository */
+    rc = funcs->C_FindObjectsInit(session, find_tmpl, find_tmpl_len);
+    if (rc != CKR_OK) {
+        testcase_fail("C_FindObjectsInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    rc = funcs->C_FindObjects(session, &h_key2, 2, &count);
```

Comment:
h_key2 is just one object handle, so you can't pass 2 as array size to C_FindObjects.

Reply:
Done.